File: 2017/lua/day19.lua
local stuff = require("stuff")

local input = {}
for line in io.lines("19.txt") do table.insert(input, stuff.str2anychartable(line)) end

local dir, row, col = {1, 0}, 1
for i, v in ipairs(input[1]) do if v ~= " " then col = i break end end

local s, count = "", 0
repeat
    row = row + dir[1]; col = col + dir[2]; count = count + 1
    local tile = input[row][col]
    if tile == "+" then
        local check = stuff.turnleft(dir)
        if input[row + check[1]][col + check[2]] ~= " " then dir = check else dir = stuff.turnright(dir) end
    elseif string.match(tile, "%a") then s = s..tile end
until input[row][col] == " "

print(string.format("Part 1: %s", s))
print(string.format("Part 2: %d", count))

File: 2017/kotlin/day19.kt
import java.io.File

fun main() {
    val input = File("../__in/19.txt").readLines()
    
    var pos = Pair(0, input[0].indexOfFirst { it != ' ' } ) // first is row index, second is column
    var dir = Pair(1, 0)            // we pretty much have regular (x, y) axes rotated 90° clockwise
    var tile: Char                  // so turnLeft / turnRight just happen to work without any changes
    var str = ""
    var step = 0
    
    do {
        pos += dir; step++
        tile = input[pos.first][pos.second]
        if (tile.isLetter()) str += tile
        else if (tile == '+') {
            with (pos + turnLeft(dir)) { dir = if (input[this.first][this.second] != ' ') turnLeft(dir) else turnRight(dir) }
        }
    } while (tile != ' ')
    
    println("Part 1: $str")
    println("Part 2: $step")
}